pkg/backend/display/rows.go

```diff
@@ -305,17 +305,41 @@ func (data *resourceRowData) ColorizedColumns() []string {
 
 	diagInfo := data.diagInfo
 
-	if done {
-		failed := data.failed || diagInfo.ErrorCount > 0
-		columns[statusColumn] = data.display.getStepDoneDescription(step, failed)
-	} else {
-		columns[statusColumn] = data.display.getStepInProgressDescription(step)
-	}
+	failed := data.failed || diagInfo.ErrorCount > 0
 
+	columns[statusColumn] = data.display.getStepStatus(step, done, failed)
 	columns[infoColumn] = data.getInfoColumn()
 	return columns
 }
 
+// addRetainStatusFlag adds a "[retain]" suffix to the input string if the resource is marked as
+// RetainOnDelete and the step will discard the resource.
+func addRetainStatusFlag(status string, step engine.StepEventMetadata) string {
+	// Guard against nil pointer dereference and return the status string as is.
+	if step.Old == nil {
+		return status
+	}
+
+	// Guard against nil pointer dereference and return the status string as is.
+	if step.Old.State == nil {
+		return status
+	}
```

Comment:
I would combine the if conditions and shorten the if statement here.
The time to read the comments here is higher than the code.
```suggestion
	// If there's no old status, we don't need to report retain.
	if step.Old == nil || step.Old.State == nil {
		return status
	}
```